### PR TITLE
Performance improvements to union

### PR DESCRIPTION
**Pull request by captncraig**

### CODE REVIEW

**Comment by captncraig** [comment]
Initial results on my box:
`BenchmarkSlowUnion	       5	 295986144 ns/op	64188939 B/op	 1001149 allocs/op`

**Comment by captncraig** [comment]
10	 192673453 ns/op	  188555 B/op	    1145 allocs/op

**Comment by captncraig** [comment]
`BenchmarkSlowUnion	      10	 194165626 ns/op	  130478 B/op	    1023 allocs/op`

**Comment by captncraig** [comment]
Didn't help this case where lengths are the same:
`BenchmarkSlowUnion	      10	 194253965 ns/op	  130488 B/op	    1024 allocs/op`

**Comment by captncraig** [comment]
`BenchmarkSlowUnion	      20	  73855974 ns/op	  130512 B/op	    1024 allocs/op`

**Comment by captncraig** [comment]
`295986144 ns/op -> 73855974 ns/op` running time is `24%` of original.

`64188939 B/op -> 130512 B/op` memory is `.2%` of original

**Comment by mjibson** [comment]
Is it possible to put the benchmarks in the commit? See: https://github.com/golang/go/commit/4bc9badd759f863f784e5f8d105a25c615324359 for how the go folks do it.

File: cmd/bosun/expr/expr.go [hunk 1]
```diff
@@ -278,35 +278,40 @@ func (e *State) union(a, b *Results, expression string) []*Union {
...
-				u.Group = rb.Group
+				group = rb.Group
 			} else {
 				continue
 			}
 			delete(am, ra)
 			delete(bm, rb)
+			u := &Union{
+				A: ra.Value,
+				B: rb.Value,
```
Comment (mjibson) [inline cmd/bosun/expr/expr.go]: `Group: group,` could be added here

File: cmd/bosun/expr/perf_test.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package expr
+
+import (
+	"bosun.org/opentsdb"
```
Comment (mjibson) [inline cmd/bosun/expr/perf_test.go]: this should be in its own block

File: cmd/bosun/expr/perf_test.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+func buildFakeResults() (ra, rb *Results) {
+	ra = &Results{}
+	rb = &Results{}
+	rand.Seed(42)
+	for i := 0; i < 50000; i++ {
+		tags := opentsdb.TagSet{}
+		tags["disk"] = fmt.Sprint("a%d", i)
+		tags["host"] = fmt.Sprint("b%d", i)
+		if i < 1000 {
+			ra.Results = append(ra.Results, &Result{Value: Number(rand.Int63()), Group: tags})
```
Comment (mjibson) [inline cmd/bosun/expr/perf_test.go]: Can we just use `0` instead of rand.Int63?

File: cmd/bosun/expr/perf_test.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		}
+		rb.Results = append(ra.Results, &Result{Value: Number(rand.Int63()), Group: tags})
+	}
+	rb.Results = rb.Results.DescByValue()
```
Comment (mjibson) [inline cmd/bosun/expr/perf_test.go]: Why do we need to sort them?

File: cmd/bosun/expr/perf_test.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"fmt"
+	"math/rand"
+	"testing"
+
+	"bosun.org/opentsdb"
+)
+
+func TestSlowUnion(t *testing.T) {
```
Comment (mjibson) [inline cmd/bosun/expr/perf_test.go]: I don't really see the point of this test.

File: cmd/bosun/expr/perf_test.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	rb.Results = rb.Results.DescByValue()
+	return ra, rb
+}
+
+func BenchmarkSlowUnion(b *testing.B) {
+	ra, rb := buildFakeResults()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		e := State{}
+		e.unjoinedOk = true
```
Comment (mjibson) [inline cmd/bosun/expr/perf_test.go]: merge with previous line

File: cmd/bosun/expr/perf_test.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	}
+	rb.Results = rb.Results.DescByValue()
+	return ra, rb
+}
+
+func BenchmarkSlowUnion(b *testing.B) {
+	ra, rb := buildFakeResults()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		e := State{}
```
Comment (mjibson) [inline cmd/bosun/expr/perf_test.go]: e is not modified during union, thus you can create this just once at the top

**Comment by kylebrandt** [comment]
Squash and merge.

**Pull request merged**